<a href="https://colab.research.google.com/github/anish1404/NLP/blob/main/Sentiment_Analysis_Sequential_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy
from keras.datasets import imdb
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import LSTM
from keras.layers.embeddings import Embedding
from keras.preprocessing import sequence
# fix random seed for reproducibility
numpy.random.seed(7)

In [ ]:
# load the dataset but only keep the top n words, zero the rest
top_words = 10000
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=top_words)

17473536/17464789 [==============================] - 0s 0us/step


<string>:6: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/keras/datasets/imdb.py:155: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  x_train, y_train = np.array(xs[:idx]), np.array(labels[:idx])
/usr/local/lib/python3.7/dist-packages/keras/datasets/imdb.py:156: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the nda

In [ ]:
X_train.shape

(25000,)

In [ ]:
X_test.shape

(25000,)

In [ ]:
X_test

array([list([1, 591, 202, 14, 31, 6, 717, 10, 10, 2, 2, 5, 4, 360, 7, 4, 177, 5760, 394, 354, 4, 123, 9, 1035, 1035, 1035, 10, 10, 13, 92, 124, 89, 488, 7944, 100, 28, 1668, 14, 31, 23, 27, 7479, 29, 220, 468, 8, 124, 14, 286, 170, 8, 157, 46, 5, 27, 239, 16, 179, 2, 38, 32, 25, 7944, 451, 202, 14, 6, 717]),
       list([1, 14, 22, 3443, 6, 176, 7, 5063, 88, 12, 2679, 23, 1310, 5, 109, 943, 4, 114, 9, 55, 606, 5, 111, 7, 4, 139, 193, 273, 23, 4, 172, 270, 11, 7216, 2, 4, 8463, 2801, 109, 1603, 21, 4, 22, 3861, 8, 6, 1193, 1330, 10, 10, 4, 105, 987, 35, 841, 2, 19, 861, 1074, 5, 1987, 2, 45, 55, 221, 15, 670, 5304, 526, 14, 1069, 4, 405, 5, 2438, 7, 27, 85, 108, 131, 4, 5045, 5304, 3884, 405, 9, 3523, 133, 5, 50, 13, 104, 51, 66, 166, 14, 22, 157, 9, 4, 530, 239, 34, 8463, 2801, 45, 407, 31, 7, 41, 3778, 105, 21, 59, 299, 12, 38, 950, 5, 4521, 15, 45, 629, 488, 2733, 127, 6, 52, 292, 17, 4, 6936, 185, 132, 1988, 5304, 1799, 488, 2693, 47, 6, 392, 173, 4, 2, 4378, 270, 2352, 4, 1500, 7, 

In [ ]:
y_test

array([0, 1, 1, ..., 0, 0, 0])

In [ ]:
# truncate and pad input sequences
max_review_length = 500
X_train = sequence.pad_sequences(X_train, maxlen=max_review_length)
X_test = sequence.pad_sequences(X_test, maxlen=max_review_length)

In [ ]:
# create the model
embedding_vecor_length = 32
model = Sequential()
model.add(Embedding(top_words, embedding_vecor_length, input_length=max_review_length))
model.add(LSTM(100))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.summary())
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=3, batch_size=64)

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 500, 32)           320000    
_________________________________________________________________
lstm (LSTM)                  (None, 100)               53200     
_________________________________________________________________
dense (Dense)                (None, 1)                 101       
Total params: 373,301
Trainable params: 373,301
Non-trainable params: 0
_________________________________________________________________
None
Epoch 1/3
391/391 [==============================] - 292s 698ms/step - loss: 0.5610 - accuracy: 0.6901 - val_loss: 0.3151 - val_accuracy: 0.8657
Epoch 2/3
391/391 [==============================] - 274s 700ms/step - loss: 0.2408 - accuracy: 0.9078 - val_loss: 0.3419 - val_accuracy: 0.8543
Epoch 3/3
391/391 [==============================] - 273s 699ms/step - loss: 0.1899 -

In [ ]:
import numpy as np

pos_cnt, neg_cnt, pos_correct, neg_correct = 0, 0, 0, 0

for x in range(len(X_test)):
    
    result = model.predict(X_test[x].reshape(1,X_test.shape[1]),batch_size=1,verbose = 2)[0]
   
    if np.argmax(result) == np.argmax(Y_test[x]):
        if np.argmax(Y_test[x]) == 0:
            neg_correct += 1
        else:
            pos_correct += 1
       
    if np.argmax(Y_test[x]) == 0:
        neg_cnt += 1
    else:
        pos_cnt += 1

print("pos_acc", pos_correct/pos_cnt*100, "%")
print("neg_acc", neg_correct/neg_cnt*100, "%")

In [ ]:
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
vocabSize = 5000
tokenizer = Tokenizer(num_words=vocabSize, split=' ')
twt = ['He is a lazy person.']
#vectorizing the tweet by the pre-fitted tokenizer instance
twt =  tokenizer.texts_to_sequences(twt)
#padding the tweet to have exactly the same shape as `embedding_2` input
twt = pad_sequences(twt, maxlen=28, dtype='int32', value=0)
print(twt)
sentiment = model.predict(twt,batch_size=1,verbose = 2)[0]
if(np.argmax(sentiment) == 0):
    print("negative")
elif (np.argmax(sentiment) == 1):
    print("positive")

[[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]
1/1 - 0s
negative


**PART 2**

In [ ]:
from keras.layers import Embedding, Flatten, Dense, LSTM, Bidirectional, Dropout, Conv1D, MaxPool1D
from keras.layers import GlobalMaxPool1D, GRU
from keras import optimizers

In [ ]:
import pandas as pd


df=pd.read_json("/content/drive/MyDrive/Sarcasm_Headlines_Dataset.json", lines=True)
df.head()

article_link  ... is_sarcastic
0  https://www.huffingtonpost.com/entry/versace-b...  ...            0
1  https://www.huffingtonpost.com/entry/roseanne-...  ...            0
2  https://local.theonion.com/mom-starting-to-fea...  ...            1
3  https://politics.theonion.com/boehner-just-wan...  ...            1
4  https://www.huffingtonpost.com/entry/jk-rowlin...  ...            0

[5 rows x 3 columns]

In [ ]:
df['is_sarcastic'].value_counts()

0    14985
1    11724
Name: is_sarcastic, dtype: int64

In [ ]:
from sklearn.model_selection import train_test_split
X=df['headline'].values
y=df['is_sarcastic'].values
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=42)

In [ ]:
X_train.shape

(18696,)

In [ ]:
y_test.shape

(8013,)

In [ ]:
vocab_size=10000
embedding_dim=16
max_length=32
trunc_type='post'
padding_type='post'
oov_tok='<oov>'
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
tokenizer= Tokenizer(num_words=vocab_size, oov_token=oov_tok)
tokenizer.fit_on_texts(X_train)
word_index=tokenizer.word_index
training_sequences=tokenizer.texts_to_sequences(X_train)
training_padded=pad_sequences(training_sequences, maxlen=max_length, padding=padding_type, truncating=trunc_type)
testing_sequences=tokenizer.texts_to_sequences(X_test)
testing_padded=pad_sequences(testing_sequences, maxlen=max_length, padding=padding_type, truncating=trunc_type)


In [ ]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Embedding, Flatten
model = Sequential()
model.add(Embedding(vocab_size, embedding_dim, input_length=max_length))
model.add(Flatten())
model.add(Dense(units=32,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=10,activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(units=1,activation='sigmoid'))
opt = tf.keras.optimizers.Adam(learning_rate=0.0001)
model.compile(loss='binary_crossentropy', optimizer=opt, metrics=['accuracy'])
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding (Embedding)        (None, 32, 16)            160000    
_________________________________________________________________
flatten (Flatten)            (None, 512)               0         
_________________________________________________________________
dense (Dense)                (None, 32)                16416     
_________________________________________________________________
dropout (Dropout)            (None, 32)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                330       
_________________________________________________________________
dropout_1 (Dropout)          (None, 10)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 1

In [ ]:
from tensorflow.keras.callbacks import EarlyStopping
early_stop = EarlyStopping(monitor='val_loss', mode='min', verbose=1, patience=5)
model.fit(x=training_padded, y=y_train, batch_size=256, epochs=100, validation_data=(testing_padded, y_test), verbose=1, callbacks=[early_stop])

Epoch 1/100
74/74 [==============================] - 1s 9ms/step - loss: 0.6894 - accuracy: 0.5438 - val_loss: 0.6873 - val_accuracy: 0.5546
Epoch 2/100
74/74 [==============================] - 0s 6ms/step - loss: 0.6855 - accuracy: 0.5618 - val_loss: 0.6846 - val_accuracy: 0.5546
Epoch 3/100
74/74 [==============================] - 1s 7ms/step - loss: 0.6841 - accuracy: 0.5623 - val_loss: 0.6823 - val_accuracy: 0.5546
Epoch 4/100
74/74 [==============================] - 1s 7ms/step - loss: 0.6795 - accuracy: 0.5637 - val_loss: 0.6777 - val_accuracy: 0.5546
Epoch 5/100
74/74 [==============================] - 1s 7ms/step - loss: 0.6716 - accuracy: 0.5696 - val_loss: 0.6677 - val_accuracy: 0.5550
Epoch 6/100
74/74 [==============================] - 1s 7ms/step - loss: 0.6587 - accuracy: 0.5871 - val_loss: 0.6505 - val_accuracy: 0.5726
Epoch 7/100
74/74 [==============================] - 1s 8ms/step - loss: 0.6387 - accuracy: 0.6172 - val_loss: 0.6256 - val_accuracy: 0.6438
Epoch 8/100
7

In [ ]:

def lstm():
    model = Sequential()
    
    model.add(Embedding(input_dim=20000, output_dim=6, input_length=40))
    
    model.add(Bidirectional(LSTM(16, return_sequences=True, recurrent_dropout=0.1, dropout=0.1)))
    
    model.add(Bidirectional(LSTM(32, recurrent_dropout=0.1, dropout=0.1)))
    
    model.add(Dense(1, activation='sigmoid'))
    
    # Compile the model
    model.compile(optimizer = 'rmsprop',
                  loss = 'binary_crossentropy',
                  metrics = ['acc'])
    
    return model

lsmod = lstm()

In [ ]:
# Train the model
lshist = lsmod.fit(training_padded, y_train,
         epochs = 10,
         batch_size = 256,
         validation_data = (testing_padded, y_test))

Epoch 1/10
74/74 [==============================] - 28s 227ms/step - loss: 0.6055 - acc: 0.6448 - val_loss: 0.6066 - val_acc: 0.7195
Epoch 2/10
74/74 [==============================] - 16s 216ms/step - loss: 0.3429 - acc: 0.8537 - val_loss: 0.3930 - val_acc: 0.8409
Epoch 3/10
74/74 [==============================] - 16s 214ms/step - loss: 0.2763 - acc: 0.8887 - val_loss: 0.6305 - val_acc: 0.7745
Epoch 4/10
74/74 [==============================] - 16s 214ms/step - loss: 0.2402 - acc: 0.9036 - val_loss: 0.4557 - val_acc: 0.8288
Epoch 5/10
74/74 [==============================] - 16s 213ms/step - loss: 0.2121 - acc: 0.9167 - val_loss: 0.3843 - val_acc: 0.8541
Epoch 6/10
74/74 [==============================] - 16s 212ms/step - loss: 0.1899 - acc: 0.9266 - val_loss: 0.4705 - val_acc: 0.8222
Epoch 7/10
74/74 [==============================] - 16s 216ms/step - loss: 0.1729 - acc: 0.9358 - val_loss: 0.4640 - val_acc: 0.8265
Epoch 8/10
74/74 [==============================] - 16s 214ms/step - 

**Using Glove Vectors**

In [ ]:
EMBEDDING_FILE = '/content/drive/MyDrive/glove.6B.200d.txt'

In [ ]:
def get_coefs(word, *arr): 
    return word, np.asarray(arr, dtype='float32')
embeddings_index = dict(get_coefs(*o.rstrip().rsplit(' ')) for o in open(EMBEDDING_FILE))

In [ ]:
max_features=20000

all_embs = np.stack(embeddings_index.values())
emb_mean,emb_std = all_embs.mean(), all_embs.std()
embed_size = all_embs.shape[1]

word_index = tokenizer.word_index
nb_words = min(max_features, len(word_index))
#change below line if computing normal stats is too slow
embedding_matrix = embedding_matrix = np.random.normal(emb_mean, emb_std, (nb_words, embed_size))
for word, i in word_index.items():
    if i >= max_features: continue
    embedding_vector = embeddings_index.get(word)
    if embedding_vector is not None: embedding_matrix[i] = embedding_vector

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2822: FutureWarning: arrays to stack must be passed as a "sequence" type such as list or tuple. Support for non-sequence iterables such as generators is deprecated as of NumPy 1.16 and will raise an error in the future.
  if self.run_code(code, result):


In [ ]:
import keras
#Defining Neural Network
model = Sequential()
#Non-trainable embeddidng layer
model.add(Embedding(nb_words, output_dim=embed_size, weights=[embedding_matrix], input_length=200, trainable=True))
#LSTM 
model.add(Bidirectional(LSTM(units=128 , recurrent_dropout = 0.5 , dropout = 0.5)))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer=keras.optimizers.Adam(lr = 0.01), loss='binary_crossentropy', metrics=['acc'])

/usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2.py:375: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  "The `lr` argument is deprecated, use `learning_rate` instead.")


In [ ]:
model.fit(training_padded, y_train,
         epochs = 10,
         batch_size = 256,
         validation_data = (testing_padded, y_test))

Epoch 1/10
74/74 [==============================] - 81s 1s/step - loss: 0.5233 - acc: 0.7448 - val_loss: 0.3425 - val_acc: 0.8520
Epoch 2/10
74/74 [==============================] - 75s 1s/step - loss: 0.2399 - acc: 0.9017 - val_loss: 0.3299 - val_acc: 0.8633
Epoch 3/10
74/74 [==============================] - 75s 1s/step - loss: 0.1380 - acc: 0.9484 - val_loss: 0.3867 - val_acc: 0.8585
Epoch 4/10
74/74 [==============================] - 74s 1s/step - loss: 0.0934 - acc: 0.9666 - val_loss: 0.4369 - val_acc: 0.8604
Epoch 5/10
74/74 [==============================] - 74s 1s/step - loss: 0.0617 - acc: 0.9782 - val_loss: 0.5537 - val_acc: 0.8561
Epoch 6/10
74/74 [==============================] - 74s 1s/step - loss: 0.0447 - acc: 0.9839 - val_loss: 0.6072 - val_acc: 0.8549
Epoch 7/10
74/74 [==============================] - 74s 1s/step - loss: 0.0402 - acc: 0.9865 - val_loss: 0.7014 - val_acc: 0.8521
Epoch 8/10
74/74 [==============================] - 74s 999ms/step - loss: 0.0287 - acc: 0

**acc: 0.9666 - val_loss: 0.4369 - val_acc: 0.8604 can be achieved**